File: packages/go/core/store/sqlite_test.go
package store

import (
	"path/filepath"
	"testing"
	"time"
)

func makeHit(overrides ...func(*HitEvent)) HitEvent {
	callerID := "key-abc"
	ua := "test/1.0"
	days := 100
	h := HitEvent{
		Ts:              time.Now().UTC(),
		Path:            "/api/v1/users",
		Method:          "GET",
		CallerID:        &callerID,
		UserAgent:       &ua,
		DaysUntilSunset: &days,
		EndpointKey:     "/api/v1/users",
		Enforced:        false,
	}
	for _, o := range overrides {
		o(&h)
	}
	return h
}

func newTestSQLite(t *testing.T) *SQLiteHitStore {
	t.Helper()
	s, err := NewSQLiteHitStore(filepath.Join(t.TempDir(), "test.db"))
	if err != nil {
		t.Fatal(err)
	}
	t.Cleanup(func() { s.Close() })
	return s
}

func TestSQLite_RecordAndRecentHits(t *testing.T) {
	s := newTestSQLite(t)
	if err := s.Record(makeHit()); err != nil {
		t.Fatal(err)
	}
	hits, err := s.RecentHits(HitQuery{Limit: 10})
	if err != nil {
		t.Fatal(err)
	}
	if len(hits) != 1 {
		t.Fatalf("expected 1 hit, got %d", len(hits))
	}
	if hits[0].Path != "/api/v1/users" {
		t.Fatalf("unexpected path: %s", hits[0].Path)
	}
}

func TestSQLite_EnforcedPersisted(t *testing.T) {
	s := newTestSQLite(t)
	s.Record(makeHit(func(h *HitEvent) { h.Enforced = true }))
	hits, _ := s.RecentHits(HitQuery{})
	if !hits[0].Enforced {
		t.Fatal("expected enforced=true")
	}
}

func TestSQLite_FilterByEndpointKey(t *testing.T) {
	s := newTestSQLite(t)
	ep2 := "/api/v1/orders/{id}"
	s.Record(makeHit())
	s.Record(makeHit(func(h *HitEvent) { h.Path = "/api/v1/orders/1"; h.EndpointKey = ep2 }))

	ep := "/api/v1/users"
	hits, _ := s.RecentHits(HitQuery{EndpointKey: &ep})
	if len(hits) != 1 || hits[0].EndpointKey != ep {
		t.Fatalf("expected 1 hit for /api/v1/users, got %d", len(hits))
	}
}

func TestSQLite_EndpointSummaries(t *testing.T) {
	s := newTestSQLite(t)
	callerA := "key-a"
	callerB := "key-b"
	ep2 := "/api/v1/orders/{id}"
	s.Record(makeHit(func(h *HitEvent) { h.CallerID = &callerA }))
	s.Record(makeHit(func(h *HitEvent) { h.CallerID = &callerB }))
	s.Record(makeHit(func(h *HitEvent) {
		h.Path = "/api/v1/orders/1"
		h.EndpointKey = ep2
		h.CallerID = &callerA
	}))

	summaries, err := s.EndpointSummaries(30)
	if err != nil {
		t.Fatal(err)
	}
	var users *EndpointSummary
	for i := range summaries {
		if summaries[i].EndpointKey == "/api/v1/users" {
			users = &summaries[i]
		}
	}
	if users == nil {
		t.Fatal("expected /api/v1/users in summaries")
	}
	if users.TotalHits != 2 {
		t.Fatalf("expected TotalHits=2, got %d", users.TotalHits)
	}
	if users.UniqueCallers != 2 {
		t.Fatalf("expected UniqueCallers=2, got %d", users.UniqueCallers)
	}
}

func TestSQLite_TotalSummary(t *testing.T) {
	s := newTestSQLite(t)
	callerA := "key-a"
	daysPos := 10
	callerB := "key-b"
	daysPast := -5
	ep2 := "/api/v1/gone"

	s.Record(makeHit(func(h *HitEvent) { h.CallerID = &callerA; h.DaysUntilSunset = &daysPos }))
	s.Record(makeHit(func(h *HitEvent) {
		h.Path = ep2; h.EndpointKey = ep2
		h.CallerID = &callerB; h.DaysUntilSunset = &daysPast; h.Enforced = true
	}))

	total, err := s.TotalSummary(30)
	if err != nil {
		t.Fatal(err)
	}
	if total.TotalHits != 2 {
		t.Fatalf("expected TotalHits=2, got %d", total.TotalHits)
	}
	if total.EndpointsWithTraffic != 2 {
		t.Fatalf("expected EndpointsWithTraffic=2, got %d", total.EndpointsWithTraffic)
	}
	if total.PastSunsetWithTraffic != 1 {
		t.Fatalf("expected PastSunsetWithTraffic=1, got %d", total.PastSunsetWithTraffic)
	}
}

func TestSQLite_OrdersNewestFirst(t *testing.T) {
	s := newTestSQLite(t)
	now := time.Now().UTC()
	s.Record(makeHit(func(h *HitEvent) { h.Ts = now.Add(-10 * time.Second) }))
	s.Record(makeHit(func(h *HitEvent) { h.Ts = now.Add(-5 * time.Second) }))

	hits, _ := s.RecentHits(HitQuery{Limit: 10})
	if len(hits) != 2 {
		t.Fatalf("expected 2 hits, got %d", len(hits))
	}
	if hits[0].Ts.Before(hits[1].Ts) {
		t.Fatal("expected newest hit first")
	}
}

File: packages/javascript/tests/core/store/sqlite.test.ts
import { describe, it, expect, beforeEach, afterEach } from 'vitest'
import { mkdtempSync, rmSync } from 'fs'
import { join } from 'path'
import { tmpdir } from 'os'
import { SQLiteHitStore } from '../../../src/core/store/sqlite'
import type { HitEvent } from '../../../src/core/store/base'

function makeHit(overrides: Partial<HitEvent> = {}): HitEvent {
  return {
    ts: new Date(),
    path: '/api/v1/users',
    method: 'GET',
    caller_id: 'key-abc',
    user_agent: 'test/1.0',
    days_until_sunset: 100,
    endpoint_key: '/api/v1/users',
    enforced: false,
    ...overrides,
  }
}

let tmpDir: string
let store: SQLiteHitStore

beforeEach(() => {
  tmpDir = mkdtempSync(join(tmpdir(), 'dusk-test-'))
  store = new SQLiteHitStore(join(tmpDir, 'test.db'))
})

afterEach(async () => {
  await store.close()
  rmSync(tmpDir, { recursive: true })
})

describe('SQLiteHitStore', () => {
  it('records and retrieves a hit', async () => {
    await store.record(makeHit())
    const hits = await store.recentHits({ limit: 10 })
    expect(hits).toHaveLength(1)
    expect(hits[0].path).toBe('/api/v1/users')
  })

  it('persists enforced=true for 410 hits', async () => {
    await store.record(makeHit({ enforced: true }))
    const hits = await store.recentHits()
    expect(hits[0].enforced).toBe(true)
  })

  it('filters hits by endpoint_key', async () => {
    await store.record(makeHit({ path: '/api/v1/users', endpoint_key: '/api/v1/users' }))
    await store.record(makeHit({ path: '/api/v1/orders/1', endpoint_key: '/api/v1/orders/{id}' }))
    const hits = await store.recentHits({ endpoint_key: '/api/v1/users' })
    expect(hits).toHaveLength(1)
    expect(hits[0].endpoint_key).toBe('/api/v1/users')
  })

  it('returns endpoint summaries with correct counts', async () => {
    await store.record(makeHit({ caller_id: 'key-a' }))
    await store.record(makeHit({ caller_id: 'key-b' }))
    await store.record(makeHit({ path: '/api/v1/orders/1', endpoint_key: '/api/v1/orders/{id}', caller_id: 'key-a' }))

    const summaries = await store.endpointSummaries()
    const users = summaries.find((s) => s.endpoint_key === '/api/v1/users')!
    expect(users.total_hits).toBe(2)
    expect(users.unique_callers).toBe(2)
  })

  it('returns total summary with past_sunset_with_traffic', async () => {
    await store.record(makeHit({ caller_id: 'key-a', days_until_sunset: 10 }))
    await store.record(makeHit({
      path: '/api/v1/gone', endpoint_key: '/api/v1/gone',
      caller_id: 'key-b', days_until_sunset: -5, enforced: true,
    }))

    const s = await store.totalSummary()
    expect(s.total_hits).toBe(2)
    expect(s.endpoints_with_traffic).toBe(2)
    expect(s.past_sunset_with_traffic).toBe(1)
  })

  it('orders recent hits newest first', async () => {
    const now = Date.now()
    await store.record(makeHit({ ts: new Date(now - 10_000) }))  // 10s ago
    await store.record(makeHit({ ts: new Date(now - 5_000) }))   // 5s ago (newer)
    const hits = await store.recentHits({ limit: 10 })
    expect(hits).toHaveLength(2)
    expect(hits[0].ts >= hits[1].ts).toBe(true)
  })
})
